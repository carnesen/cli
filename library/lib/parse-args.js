"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseArgs = void 0;
const cli_usage_error_1 = require("./cli-usage-error");
/**
 * Calls the parse method of an ArgParser
 *
 * @param argParser - An ArgParser
 * @param args - An array of string arguments
 * @param separator - A string that will be prepended to error messages (e.g. "--users")
 * @returns The result of parse
 */
async function parseArgs(argParser, args, separator, commandStack) {
    const { required, placeholder, parse } = argParser;
    const fullContext = [separator, placeholder].filter((str) => Boolean(str)).join(' ');
    const prefix = fullContext ? `${fullContext} : ` : '';
    if (required && (!args || args.length === 0)) {
        throw new cli_usage_error_1.CliUsageError(`${prefix}argument is required`, commandStack);
    }
    try {
        return await parse(args);
    }
    catch (exception) {
        if (exception && typeof exception.message === 'string') {
            exception.message = `${prefix}${exception.message}`;
        }
        if (exception && exception.code === cli_usage_error_1.CLI_USAGE_ERROR) {
            exception.commandStack = commandStack;
        }
        throw exception;
    }
}
exports.parseArgs = parseArgs;
//# sourceMappingURL=parse-args.js.map