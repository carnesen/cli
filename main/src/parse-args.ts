import { Leaf } from './cli-node';
import { CliUsageError, CLI_USAGE_ERROR } from './cli-usage-error';
import { AnyArgParser } from './cli-arg-parser';

/**
 * Calls the parse method of an ArgParser
 *
 * @param argParser - An ArgParser
 * @param args - An array of string arguments
 * @param separator - A string that will be prepended to error messages (e.g. "--users")
 * @returns The result of parse
 */
export async function parseArgs(
  argParser: AnyArgParser,
  args: string[] | undefined,
  separator: string | undefined,
  locationInCommandTree: Leaf,
): Promise<any> {
  const { required, placeholder, parse } = argParser;
  const fullContext = [separator, placeholder].filter((str) => Boolean(str)).join(' ');
  const prefix = fullContext ? `${fullContext} : ` : '';
  if (required && (!args || args.length === 0)) {
    throw new CliUsageError(`${prefix}argument is required`, locationInCommandTree);
  }
  try {
    return await parse(args);
  } catch (exception) {
    if (exception && typeof exception.message === 'string') {
      exception.message = `${prefix}${exception.message}`;
    }
    if (exception && exception.code === CLI_USAGE_ERROR) {
      exception.locationInCommandTree = locationInCommandTree;
    }
    throw exception;
  }
}
