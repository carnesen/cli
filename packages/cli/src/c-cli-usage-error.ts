import { CCliTree } from './c-cli-tree';

/** Thrown to print usage documentation and an error message */
export class CCliUsageError extends Error {
	/**
	 * @param message If provided, the {@link CCli} runner will also print
	 * "Error: \<your message\>" when it encounters this error
	 * @param tree Used internally for constructing the command-line usage string
	 * */
	public constructor(public readonly message = '', public tree?: CCliTree) {
		super(message);
	}
}
