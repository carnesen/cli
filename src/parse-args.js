'use strict'

const {assertArray} = require('@carnesen/util')

module.exports = function parseArgs (args) {
  assertArray(args, 'args')
  const parameterArgs = []
  const commandNameArgs = []
  let currentArray = commandNameArgs
  args.forEach(function (arg) {
    const matches = arg.match(/^--(.*)/)
    if (matches) {
      const name = matches[1]
      const existingCommandArg = parameterArgs.find(function (commandArg) {
        return commandArg.name === name
      })
      if (existingCommandArg) {
        throw new Error('Each parameter can be provided at most one time')
      } else {
        currentArray = []
        parameterArgs.push({name, rawValues: currentArray})
      }
    } else {
      currentArray.push(arg)
    }
  })
  return {
    parameterArgs,
    commandNameArgs,
  }
}
