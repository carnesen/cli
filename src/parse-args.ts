import { AnyArgParser, LeafStack } from './types';
import { CliUsageError, CLI_USAGE_ERROR } from './cli-usage-error';

/**
 * Calls the parse method of an ArgParser
 *
 * @param argParser - An ArgParser
 * @param args - An array of string arguments
 * @param separator - A string that will be prepended to error messages (e.g. "--users")
 * @returns The result of parse
 */
export async function parseArgs(
  argParser: AnyArgParser,
  args: string[] | undefined,
  separator: string | undefined,
  leafStack: LeafStack,
): Promise<any> {
  const { required, placeholder, parse } = argParser;
  const fullContext = [separator, placeholder].filter((str) => Boolean(str)).join(' ');
  const prefix = fullContext ? `${fullContext} : ` : '';
  if (required && (!args || args.length === 0)) {
    throw new CliUsageError(`${prefix}argument is required`, leafStack);
  }
  try {
    return await parse(args);
  } catch (exception) {
    if (exception && typeof exception.message === 'string') {
      exception.message = `${prefix}${exception.message}`;
    }
    if (exception?.code === CLI_USAGE_ERROR) {
      exception.commandStack = leafStack;
    }
    throw exception;
  }
}
