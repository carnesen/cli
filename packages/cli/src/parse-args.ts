import { CCliUsageError } from './c-cli-usage-error';
import { CCliArgGroup } from './c-cli-arg-group';

/** Call the parse method of a {@link CCliArgGroup}
 * @param argGroup - An ArgGroup
 * @param args - An array of string arguments
 * @param separator - A string that will be prepended to error messages (e.g. "--users")
 * @returns The result of parse */
export async function parseArgs(
	argGroup: CCliArgGroup,
	args: string[] | undefined,
	separator: string | undefined,
): Promise<any> {
	const fullContext = [separator, argGroup.placeholder]
		.filter((str) => Boolean(str))
		.join(' ');
	const prefix = fullContext ? `${fullContext} : ` : '';
	if (!argGroup.optional && (!args || args.length === 0)) {
		throw new CCliUsageError(`${prefix}argument is not optional`);
	}
	try {
		return await argGroup.parse(args);
	} catch (exception: any) {
		if (exception && typeof exception.message === 'string') {
			exception.message = `${prefix}${exception.message}`;
		}
		throw exception;
	}
}
