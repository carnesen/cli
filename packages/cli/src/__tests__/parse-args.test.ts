import { runAndCatch } from '@carnesen/run-and-catch';

import { parseArgs } from '../parse-args';
import {
	dummyOptionalArgGroup,
	dummyNonOptionalArgGroup,
	DUMMY_ARG_GROUP_THROWN_INTENTIONALLY,
	DUMMY_ARG_GROUP_THROW,
	DUMMY_ARG_GROUP_THROW_NON_TRUTHY,
} from '../dummy-arg-groups';
import { CCliUsageError } from '../c-cli-usage-error';

describe(parseArgs.name, () => {
	it(`returns parse(args) if an args with length >= 1 is passed`, async () => {
		const args = ['foo'];
		expect(await parseArgs(dummyOptionalArgGroup, args, undefined)).toBe(
			dummyOptionalArgGroup.parse(args),
		);
		expect(await parseArgs(dummyNonOptionalArgGroup, args, undefined)).toBe(
			dummyNonOptionalArgGroup.parse(args),
		);
	});

	it(`if not optional, returns parse(args) if args is an empty array or undefined`, async () => {
		expect(await parseArgs(dummyOptionalArgGroup, [], undefined)).toBe(
			dummyOptionalArgGroup.parse([]),
		);
		expect(await parseArgs(dummyOptionalArgGroup, undefined, undefined)).toBe(
			dummyOptionalArgGroup.parse(undefined),
		);
	});

	it(`throws usage error "argument is not optional" if args is an empty array or undefined`, async () => {
		for (const args of [undefined, [] as string[]]) {
			const exception = await runAndCatch(
				parseArgs,
				dummyNonOptionalArgGroup,
				args,
				undefined,
			);
			expect(exception).toBeInstanceOf(CCliUsageError);
			expect(exception.message).toMatch(/argument is not optional/i);
			expect(exception.message).toMatch(
				dummyNonOptionalArgGroup.placeholder || '',
			);
		}
	});

	it(`if throws "argument is not optional", expect message to match snapshot`, async () => {
		const exception = await runAndCatch(
			parseArgs,
			dummyNonOptionalArgGroup,
			undefined,
			undefined,
		);
		expect(exception.message).toMatch('argument is not optional');
		expect(exception.message).toMatchSnapshot();
	});

	it(`if throws "argument is not optional" with context, expect message to match snapshot`, async () => {
		const exception = await runAndCatch(
			parseArgs,
			dummyNonOptionalArgGroup,
			undefined,
			'context',
		);
		expect(exception.message).toMatch('argument is not optional');
		expect(exception.message).toMatchSnapshot();
	});

	it(`throws if parse does with a context/placeholder enhanced message`, async () => {
		const exception = await runAndCatch(
			parseArgs,
			dummyOptionalArgGroup,
			[DUMMY_ARG_GROUP_THROW],
			undefined,
		);
		expect(exception.message).toMatch(DUMMY_ARG_GROUP_THROWN_INTENTIONALLY);
		expect(exception.message).toMatch(dummyOptionalArgGroup.placeholder || '');
		expect(exception.message).toMatchSnapshot();
	});

	it(`just re-throws exception if parse throws a non-truthy exception`, async () => {
		const exception = await runAndCatch(
			parseArgs,
			dummyOptionalArgGroup,
			[DUMMY_ARG_GROUP_THROW_NON_TRUTHY],
			undefined,
		);
		expect(exception).not.toBeTruthy();
	});
});
