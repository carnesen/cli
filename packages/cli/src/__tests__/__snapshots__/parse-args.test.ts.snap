// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseArgs if throws "argument is not optional" with context, expect message to match snapshot 1`] = `"context <foo> : argument is not optional"`;

exports[`parseArgs if throws "argument is not optional", expect message to match snapshot 1`] = `"<foo> : argument is not optional"`;

exports[`parseArgs throws if parse does with a context/placeholder enhanced message 1`] = `"<val> : thrown intentionally"`;
